spec/client/ics-028-wasm-client/README.md

```diff
@@ -0,0 +1,451 @@
+---
+ics: 28
+title: WASM Client
+stage: draft
+category: IBC/TAO
+kind: instantiation
+implements: 2
+author: Parth Desai <parth@chorus.one>
+created: 2020-10-13
+modified: 2020-10-13
+---
+
+## Synopsis
+
+This specification document describes an interface to a client (verification algorithm) stored as a wasm bytecode for a blockchain.
+
+### Motivation
+
+WASM based light clients are decoupled from SDK source code, which enables one to upgrade existing light client or add support for other blockchain without modifying SDK source code.
+
+### Definitions
+
+Functions & terms are as defined in [ICS 2](../ics-002-client-semantics).
+
+`currentTimestamp` is as defined in [ICS 24](../ics-024-host-requirements).
+
+`WASM Client Code` refers to wasm bytecode stored in the client store, which provides target blockchain specific implementation of [ICS 2](../ics-002-client-semantics).
+
+`WASM Client` refers to a particular instance of `WASM Client Code` defined as a tuple `(WASM Client Code, ClientID)`.
+
+`WASM VM` refers to a virtual machine capable of executing valid wasm bytecode.
+
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification depends on correctness of the `WASM Client code` in context of consensus algorithm of its target `blockchain`, as well as correct instantiation of `WASM Client`.
+
+### Client state
+
+The wasm client state tracks location of the wasm bytecode via `codeId`. Binary data represented by `data` field is opaque and only interpreted by the WASM Client Code. `type` represents client type.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ClientState {
```

Comment:
why can't this whole type be up to the WASM client code

Reply:
@cwgoes That was the initial idea. 

`ClientState`, `ConsensusState` and `Misbehaviour` interfaces  have some functions like `GetProofSpec`, `IsFrozen`, `Validate` which do not give access to store and/or are being called during tx creation on client side. In both cases we do not have access to the WASM VM. But, we still have to return valid answer as per semantics of that function and to do that we need to have some structure instead of opaque blob of byte only understood by WASM code.

Reply:
This is a good point - if the spec defines data structures that must be validated without any access to storage of the VM, then these types must be understood by the Go code.

However, this is in the spec level and I wonder if "In both cases we do not have access to the WASM VM." is obvious in the spec or an implementation detail we could work to improve.

Could you link the functions that require inspection of these types, so they cannot be defined in Wasm. It might make sense to revisit those definitions with @cwgoes and see if we can make them more flexible.

Reply:
Hmm, that's an excellent point, but I think we may want to alter the structure of IBC or whatever code is involved here so that the VM _can_ be accessed - sounds like we're on the same page @ethanfrey.

Reply:
Maybe @colin-axner or @AdityaSripal can offer some more context here and/or thoughts on whether we should change the interface.

Reply:
Thanks for the detailed list of those methods.

`ProofSpecs` is static over the lifetime of the client and there may be somewhere to include this in code not in state.

`LatestHeight` obviously requires a query to the KV store. How does the current Go code handle returning this without storage access?

Reply:
Thanks, @ParthDesai and @mkaczanowski for the explanations. In principle, I'm ok with breaking any Go APIs that are not expected by counterparty chains or relayers if it will help make WASM and future client code more clean.

`ClientType() string`: Agreed that you should just be returning `10-wasm` here.

`Validate() error`: Since this is getting used in messages, it has to be stateless. Any additional checks you want to do with state must be done elsewhere. We could rename this to `ValidateBasic` to emphasize this.

`ZeroCustomFields`: This is just used for upgrades, we can provide store here if necessary.

---

### Interfaces that may be expected by counterparty chains

`GetProofSpecs`: The proof specs get used in the client `Verify` functions. This of course doesn't require an external interface, in face we just use the clientstate field directly. However, it also gets used on connection handshake to ensure that the counterparty client is using the correct proof specs: [here](https://github.com/cosmos/ibc-go/blob/2e95805e2d/modules/core/02-client/keeper/keeper.go#L301). Currently, it is not using the interface itself but rather getting the `ProofSpecs` field directly because we are making an assumption that all counterparty clients for an SDK chain are `07-tendermint` clients. However, once `WASM` client is introduced we need to remove that assumption since we could have a wasm smart contract client tracking the SDK chain (this could be better than 07-tendermint since it may have easier upgrade logic). So once that happens, we will have to rely on the interface itself. We cannot give you access to the store here, since this is getting called by the **counterparty** chain, which does not contain the store for your light client. You will need to find a way to provide this information in the client state struct itself.

`GetLatestHeight() Height`: Similar to the `ProofSpec` case, we check that the latest height of the counterparty client is not greater than our height during connection handshake [here](https://github.com/cosmos/ibc-go/blob/2e95805e2d/modules/core/02-client/keeper/keeper.go#L295). Again we currently use the field directly, but will have to switch to using an interface. 

We don't strictly require the interface, seeing as we could just execute separate `ValidateSelfClient` logic for wasm/go clients, however the wasm client must still have proof specs and latest height directly retrievable from its client state in order to perform all necessary checks on connection handshake.

Reply:
> LatestHeight obviously requires a query to the KV store. How does the current Go code handle returning this without storage access?

We "cache" the latest height directly into the client state on update: https://github.com/cosmos/ibc-go/blob/2e95805e2d/modules/light-clients/07-tendermint/types/update.go#L251

Reply:
Thanks, @AdityaSripal for such a detailed response

> however the wasm client must still have proof specs and latest height directly retrievable from its client state in order to perform all necessary checks on connection handshake.

We'll update the ICS with:
```
interface ClientState {
  codeId: []byte
  data: []byte
  
  latestHeight: Height
  proofSpecs: ProofSpec
}
```

> ZeroCustomFields: This is just used for upgrades, we can provide a store here if necessary.

Since the method intends to reset some fields in the ClientState, I don't think we need storage access. However, possibly changing the interface to allow returning the `error` would be helpful. Please see the initial implementation:
https://github.com/ChorusOne/ibc-go/blob/ics_28_wasm_client/modules/light-clients/10-wasm/types/client_state.go#L228

what do you think?

Reply:
updated the `ClientState` interface [here](https://github.com/ParthDesai/ibc/commit/d7c33fc5c01ff4c07000f89ff9da04c31c27ed23#diff-081ac6bcbed0e39a406e9c83aa79a058605a646a4047d5be67cb84dd9b83877bL48)